### ISSUE TITLE
Enable stricter validation of plugins by default for validatePlugins task

### ISSUE THREAD

**Issue by 3flex** [gradle/gradle#22600]
Gradle should strictly validate plugins written by plugin authors who have applied `java-gradle-plugin` plugin to their projects.

### Expected Behavior
`enableStricterValidation` is set to `true` by default for `ValidatePlugins` tasks.

### Current Behavior
Plugin authors have to know that this option exists, and enable it, to get all validations to run on their plugins. This means there are many plugins in the wild that do not get validated by default.

### Context
I use Gradle plugins. I like it when they behave :)

**Comment by eskatos** [comment]
This feature request is in the backlog of the relevant team but this area of Gradle is currently not a focus one so it might take a while before it gets implemented.

****

Work validation also happens at runtime. Both plugin build-time validation and runtime validation should be considered as part of this issue.

**Comment by 3flex** [comment]
I'd be happy to create a PR? It seems a simple one to implement.

**Comment by TWiStErRob** [comment]
@akshayabd this might be an answer to my question about checking code.

**Comment by lptr** [comment]
Another option could be to remove stricter validation as a separate option, and make it the only option. The only difference this would make is that inputs on tasks not marked as `@CacheableTask` would still be validated for the presence of path normalization.

**Comment by 3flex** [comment]
Nag when false in 8.x, change default to true in 9, remove in 10?

**Comment by liutikas** [comment]
Are we still considering this for Gradle 9.0?

**Comment by 3flex** [comment]
> Nag when false in 8.x, change default to true in 9, remove in 10?

@eskatos I saw you gave this a thumbs up, but @ov7 added the issue to the 9.0 RC1 milestone. Wouldn't the deprecation nagging have to be added in 8.14 (or 8.15, if there is one)? Or are you thinking of going a different direction that doesn't require any 8.x changes?

Sorry to @ but it's time sensitive if changes have to be made to 8.x and 8.14 is the last planned 8.x release.

**Comment by lptr** [comment]
Right now in non-strict mode we are not reporting problems that affect only cacheable tasks at all.

We could make this change in two steps:

1) Start by reporting these problems in 9.1, but make sure we only report them as `WARNING`s; i.e. in `DefaultTypeValidationContext:

    ```java
    @Override
    protected void recordProblem(InternalProblem problem) {
        if (onlyAffectsCacheableWork(problem.getDefinition().getId()) && !reportCacheabilityProblems) {
            problem = problem.toBuilder().severity(Severity.WARNING).build();
        }
        problems.add(problem);
    }
    ```

   We'd need to also deprecate setting reporting to non-strict explicitly.

2) In Gradle 10.0 report them as their base severity, and remove the non-strict option altogether.

**Comment by liutikas** [comment]
I strongly suggest we just change the default. Every suggestion from this task is critical to have a modern plugin and plugin authors are the ones that need to do the work. Waiting for 10 is simply too slow.

**Comment by ZacSweers** [comment]
+1. No reason to wait till 10 when this is configurable behavior that can be just disabled if it's undesirable

**Comment by pablobaxter** [comment]
Absolutely make this default for Gradle 9! I don't understand why waiting for 10 is preferable.

**Comment by yogurtearl** [comment]
Agreed, this is important and should be enabled in 9.x and not wait until 10.x.

**Comment by autonomousapps** [comment]
Agreed with the chorus above. We've been telling people to write cacheable tasks for years, and it feels really bad to leave an important validation on the table for even longer, especially when it could be disabled if necessary.

**Comment by joshfriend** [comment]
About a month ago we discovered a task in our build that was generating protos classes which should have been cached, but was missing `@CacheableTask` and would have been caught with strict validation enabled. Today I enabled strict validation and found at least 2 other tasks which were not marked cacheable that will likely benefit from cache.

Additionally, I fixed 33 other tasks by adding explicit cacheability annotations and fixing path normalization issues. That's probably more work than other plugin libraries would need to do to adopt this change, and it was not really that difficult to do (took a couple hours).

This check is _trivially easy to disable_ if enabled by default in the next _major version_ of Gradle. If the providers api migration was in serious consideration for Gradle 9, this comparatively simple one should be a no brainer to include and should not be delayed.

**Comment by martinbonnin** [comment]
[Enabling the flag in Apollo](https://github.com/apollographql/apollo-kotlin/pull/6500) found 1 missing `@PathSensitive` and 3 missing `@DisableCachingByDefault` annotations. 

Also without this issue, I would never have discovered the existence of the flag. Might be worth putting it in the top-level `gradlePlugins {}` extension for better discoverability?

**Comment by liutikas** [comment]
Given no updates on this, does that mean that Grade will not do this default change until Gradle 10.0.0 a year from now?

**Comment by eskatos** [comment]
This is moving forward, but only for builds of plugins that are published at first, as part of:
* https://github.com/gradle/gradle/issues/35497
